public/app/features/explore/Logs/Logs.tsx

```diff
@@ -248,84 +239,103 @@ class UnthemedLogs extends PureComponent<Props, State> {
-          },
-          ref: null,
-          color: LogLevelColor[level.logLevel],
-        });
+        const allLevelsSelected = hiddenLogLevels.length === 0;
+        const currentLevelSelected = !hiddenLogLevels.find((hiddenLevel) => hiddenLevel === level.levelStr);
+        if (register) {
+          register({
+            title: level.levelStr,
+            icon: 'gf-logs',
+            panelId: 'Logs',
+            level: 'child',
+            type: 'filter',
+            highlight: currentLevelSelected && !allLevelsSelected,
+            onClick: (e: React.MouseEvent) => {
+              toggleLegendRef.current?.(level.levelStr, mapMouseEventToMode(e));
+            },
+            ref: null,
+            color: LogLevelColor[level.logLevel],
+          });
+        }
       });
     }
-  };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [logsVolumeData?.data, hiddenLogLevels]);
 
-  updatePanelState = (logsPanelState: Partial<ExploreLogsPanelState>) => {
-    const state: ExploreItemState | undefined = getState().explore.panes[this.props.exploreId];
-    if (state?.panelsState) {
-      dispatch(
-        changePanelState(this.props.exploreId, 'logs', {
-          ...state.panelsState.logs,
-          columns: logsPanelState.columns ?? this.props.panelState?.logs?.columns,
-          visualisationType: logsPanelState.visualisationType ?? this.state.visualisationType,
-          labelFieldName: logsPanelState.labelFieldName,
-          refId: logsPanelState.refId ?? this.props.panelState?.logs?.refId,
-        })
-      );
-    }
-  };
+  useEffect(() => {
+    return () => {
+      if (flipOrderTimer) {
+        window.clearTimeout(flipOrderTimer);
+      }
 
-  getNumberOfLogVolumes() {
-    const data = this.props.logsVolumeData?.data.filter(
-      (frame: DataFrame) => frame.meta?.dataTopic !== DataTopic.Annotations
-    );
-    const grouped = groupBy(data, 'meta.custom.datasourceName');
-    const numberOfLogVolumes = Object.keys(grouped).length;
-    return numberOfLogVolumes;
-  }
+      if (cancelFlippingTimer) {
+        window.clearTimeout(cancelFlippingTimer);
+      }
 
-  componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<State>): void {
-    if (this.props.loading && !prevProps.loading && this.props.panelState?.logs?.id) {
+      // If we're unmounting logs (e.g. switching to another datasource), we need to remove the table specific panel state, otherwise it will persist in the explore url
+      if (panelState?.logs?.columns || panelState?.logs?.refId || panelState?.logs?.labelFieldName) {
+        dispatch(
+          changePanelState(exploreId, 'logs', {
+            ...panelState?.logs,
+            columns: undefined,
+            visualisationType: visualisationType,
+            labelFieldName: undefined,
+            refId: undefined,
+          })
+        );
+      }
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Disabling exhaustive deps should be an exceptional measure of last resort, so in general in this PR I would ask you to not use it.

For this particular case, the clean up function will be created and returned with values that might not be valid or present when it's executed, specially those pointing to timers.

Reply:
Good catch! We may need to use `useRef` to store latest `panelState.logs` and call it only on unmount. In long-run this code looks like should not be here at all and Explore should clean up panel state for each panel automatically 🤔

Reply:
Fixed in effed375e1bfadcaec7ed926fed926d3445f8e05 + a75b5c80be0994f09e3ea7987d867acc65b52633.